[docs] Update Push notifications and expo-notifications docs (#20270)
expo/expo: docs/pages/push-notifications/sending-notifications-custom.mdx

---
title: Send notifications with FCM & APNs
hideFromSearch: true
---

Communicating directly with FCM and APNs is much more complicated than sending notifications through [Expo's push notification service]/push-notifications/sending-notifications/). You should only use this feature if you're prepared to undertake that complexity. Here are a few things you'll have to handle yourself if you choose to write your server for FCM and APNs:

- Differentiating between native Android & iOS device tokens on your backend
- Twice the amount of backend code to write and maintain (code for communicating with FCM, and then code for communicating with APNs)
- Fetching responses from FCM and APNs to check if your notification went through, error handling, credentials management

> Sometimes you need finer-grained control over your notifications, so communicating directly with FCM and APNs is necessary. The Expo platform does not lock you into using Expo's application services, and the `expo-notifications` API is push-service agnostic.

Reviewer: ```suggestion
You may need finer-grained control over your notifications, so communicating directly with FCM and APNs may be necessary. The Expo platform does not lock you into using Expo's application services, and the `expo-notifications` API is push-service agnostic.
```

In my opinion, everything above this paragraph is not needed. I keep on reading (over and over) how Expo's push notification service is easy and removes complexity. That seems relevant for an overview page but I think we can remove that from pages like this one.

## How do I write my own FCM & APNs servers

Before communicating directly with FCM & APNs, there is one client-side change you'll need to make in your app. When using Expo's notification service, you collect the `ExponentPushToken` with [`getExpoPushTokenAsync`](/versions/latest/sdk/notifications/#getexpopushtokenasyncoptions-expotokenoptions-expopushtoken). Now that you're not using Expo's notification service, you'll need to collect the native device token instead with [`getDevicePushTokenAsync`](/versions/latest/sdk/notifications/#getdevicepushtokenasync-devicepushtoken).

```diff
import * as Notifications from 'expo-notifications';
...
- const token = (await Notifications.getExpoPushTokenAsync()).data;
+ const token = (await Notifications.getDevicePushTokenAsync()).data;
// send token to your server
```

After getting the native device token, you can start implementing the servers. Below are some minimal examples of communicating with FCM and APNs:

## FCM server

> This documentation is based on [Google's documentation](https://firebase.google.com/docs/cloud-messaging/http-server-ref), and this section covers the basics to get you started.

> Communicating with FCM is as simple as sending a POST request. However, before sending or receiving any notifications, you'll need to follow the steps to [configure FCM](/push-notifications/push-notifications-setup/#android) to configure FCM and get your `FCM-SERVER-KEY`.

Reviewer: ```suggestion
Communicating with FCM is done by sending a POST request. However, before sending or receiving any notifications, you'll need to follow the steps to [configure FCM](/push-notifications/push-notifications-setup/#android) to configure FCM and get your `FCM-SERVER-KEY`.
```
We say "is as simple as a POST request" and then immediately explain that it's harder than that 😂 --Just taking out that statement here.
